### PR TITLE
Add: Tutorial for credentials provider using custom backend

### PR DESCRIPTION
**Pull request by twilink**
**What**:

An in-depth tutorial of how to use credentials provider with a custom backend including how to display error messages received from server (400-500 codes) on your custom login page

**Why**:

Was unable to find a detailed example of using the credentials provider with a custom backend and error handling. Also saw questions in stackoverflow (https://stackoverflow.com/questions/65297986/how-to-ensure-auth-with-axios-in-custom-backend-with-credentials-provider/65845618#65845618) and nextauth discussions requesting the same (https://github.com/nextauthjs/next-auth/discussions/661#discussioncomment-154364)

Hope it helps!

**Checklist**:

- [x] Documentation
- [ ] Tests N/A
- [x] Ready to be merged

### CODE REVIEW

**Review by balazsorban44 — CHANGES_REQUESTED** [review]
Thank you! I'll look at it in-depth sometime later. Could you in the meantime remove the `Promise.resolve` and `Promise.reject` wrappers in your code snippets? They are unnecessary, and only cause confusion amongst users.

Also, feel free to disagree, but could you change property arrow functions to methods like so:
this:
```js
jwt: async (token, user) => {
  if (user) {
    token.accessToken = user?.data.token
  }
  return token
},
```
to this (no need for optional chaining on user, you already checked in the if block if user is defined):
```js
async jwt(token, user) {
  if (user) {
    token.accessToken = user.data.token
  }
  return token
}
``` 
(Look for similarity defined property functions and replace those as well)

This might only be my personal preference, but then it would also better match with our code examples in the docs.

With our `canary` release (hopefully soon to be `latest`), you can also simplify the following:
```js
const options = {
  providers,
  callbacks
  pages: {
    error: '/login' // Changing the error redirect page to our custom login page
  }
}

export default (req, res) => NextAuth(req, res, options)
```
to
```js
export default NextAuth({
  providers,
  callbacks
  pages: {
    error: '/login' // Changing the error redirect page to our custom login page
  }
})
```

You don't have to change this, since it is not yet in the stable release, but please keep an eye on when this gets to stable and change your article accordingly, so that we know the tutorials out there are up-to-date 🙂


Again, I have to read through the tutorial again, these small improvements/changes were from the top of my head only.

Nice work either way!

**Comment by twilink** [comment]
Thank you for the comment! 🙂

I have updated the docs to reflect the changes you suggested and will keep an eye out for the stable release to update the export function.

Please let me know if any other changes are required. Appreciate the detailed review! 🙂

**Review by balazsorban44 — COMMENTED** [review]
TL;DR

**Since we already have an issue/request to better handling this problem, I am not going to accept this PR, but I appreciate the effort anyway!**

Some further tips/suggestions for your article though:
1. `pages/api/[..next-auth].js` -> `pages/api/auth/[...next-auth].js`
2. could you maybe remove `axios` to keep the example code as simple as possible? The less external libraries, the better. A `fetch` polyfill is included in Next.js already, and it is barely more code.
3. The authorize callback could be as simple as
```js
async authorize(user) {
  return fetch("https://myapi.com/login", {
      headers: { "Content-Type": "application/json", method: "POST" },
      body: JSON.stringify({user})
  })
}
```
4. the `signIn` callback is unnecessary in the form you defined it, just remove it.
5. The `session` callback also needs an if block to check if `session` already have the accessToken, I think. Could you check if your example code actually returns an `accessToken` on `/api/auth/session`?
6. Would be nice to show the whole custom login page code, not just the `handleLogin` function.


Further explanation to the first sentence: To be honest, I think the best way would be if we just let the `signIn` function return something (like errors, and success), rather than requiring users to create custom pages and all that fuss. (see #517, #989)

**Comment by twilink** [comment]
> TL;DR
> 
> **Since we already have an issue/request to better handling this problem, I am not going to accept this PR, but I appreciate the effort anyway!**
> 
> Some further tips/suggestions for your article though:
> 
>     1. `pages/api/[..next-auth].js` -> `pages/api/auth/[...next-auth].js`
> 
>     2. could you maybe remove `axios` to keep the example code as simple as possible? The less external libraries, the better. A `fetch` polyfill is included in Next.js already, and it is barely more code.
> 
>     3. The authorize callback could be as simple as
> 
> 
> ```js
> async authorize(user) {
>   return fetch("https://myapi.com/login", {
>       headers: { "Content-Type": "application/json", method: "POST" },
>       body: JSON.stringify({user})
>   })
> }
> ```
> 
>     1. the `signIn` callback is unnecessary in the form you defined it, just remove it.
> 
>     2. The `session` callback also needs an if block to check if `session` already have the accessToken, I think. Could you check if your example code actually returns an `accessToken` on `/api/auth/session`?
> 
>     3. Would be nice to show the whole custom login page code, not just the `handleLogin` function.
> 
> 
> Further explanation to the first sentence: To be honest, I think the best way would be if we just let the `signIn` function return something (like errors, and success), rather than requiring users to create custom pages and all that fuss. (see #517, #989)

Okay! Yeah I agree having the `signIn` function return errors would be the best way to handle these scenarios. Looking forward to having that option in future release.

Thank you for taking the time to go through the article and for the suggestions. I shall update the article accordingly.
